### ISSUE TITLE
two virtual-object GC problems

### ISSUE THREAD

**Issue by warner** [Agoric/Agoric#5044]
@FUDCo and I noticed two problems in the virtual-object GC code today:

## Collections vs Remotables in values

`collectionManager.js` has a function named `deleteInternal()` used to delete an entry (key plus value) from a map-like collection. The value may contain some arbitrary number of vref-indexed objects. When the entry goes away, we need to decref each of those slots, by calling `vrm.removeReachableVref(vref)` on them. There is a loop that looks like:
```js
  value.slots.map(vrm.removeReachableVref);
```

The problem is that `vrm.removeReachableVref(vref)` has a boolean return value `doMoreGC`, which is `true` when the vref in question pointed to a Remotable (in RAM), and that Remotable's refcount just dropped to 0. In that case, we're going to be freeing some RAM, and that RAM might have captured/closed-over some other RAM presences/representatives/remotables. So if that ever happens, for *any* slot in this `value`, we need to inform our caller, so they can perform a new `gcAndFinalize` to let those new finalizers get noticed.

The loop was ignoring the return value. Instead, we need something like
```js
value.slots.forEach(vref => { doMoreGC = vrm.removeReachableVref(vref) || doMoreGC; } );
```

To test this, we'll need a collection with a string-keyed entry whose value contains a `[remotable, presence1]`. The Remotable's definition should closes over a `presence2`. Nothing else should keep either Presence alive. Allow everything to settle, then delete the collection entry (without causing anything to be deleted from RAM: no Presences or Remotables or Representatives should go out of scope on the delivery that does `collection.delete('stringkey')`).

On the next `bringOutYourDead`, the Remotable should be collected along with the vref for `presence2`, leaving `presence1` unreferenced. However the bug will inhibit a second call to `gcAndFinalize`, so that BOYD will *not* observe the presence1 finalizer run. The test program should be able to observe the presence-2 vref being missing from the c-list, but the presence1-vref still being present. A second BOYD (e.g. any random message, assuming the `reapInterval` is set to 1) will flush out presence-1 vref.

Once the bug is fixed, both presence-1 and presence-2 vrefs should be removed from the c-list during the first message, and it should not require a second message to get both.

## Virtual-Object Properties vs doMoreGC

In `virtualObjectManager.js`, the `deleteStoredVO()` function does a similar call to `vrm.removeReachableVref()` which *does* preserve the "do more" flag:

```js
        for (const propValue of Object.values(rawState)) {
          propValue.slots.map(
            vref => (doMoreGC = doMoreGC || vrm.removeReachableVref(vref)),
          );
        }
```

The problem is that if one call sets `doMoreGC`, subsequent calls will be bypassed by the short-circuit operator.

The fix is to swap the arguments:
```js
            vref => (doMoreGC = vrm.removeReachableVref(vref) || doMoreGC),
```

To test this one, we need a virtual object with one property. The value of this property should have two slots, like `state.x = [remotable, presence]`. The first slot should be a Remotable, so `doMoreGC` is set to `true`. The bug will cause the second slot to not be removed. By using a presence, the test program can inspect the vat's c-list to see whether it was dropped or not.

```js
// target vat
export function buildRootObject() {
  let foo;
  const makeFoo = defineKind('foo', (rem, pres) => { x: [rem, pres] }, {});
  return Far('root', {
    create(pres) { foo = makeFoo(Far('rem', {}), pres); },
    forget() { foo = 0; },
  });
}
```

```js
// bootstrap vat
export functionbuildRootObject() {
  const sensor = Far('sensor', {});
  let target;
  return Far('root', {
    bootstrap(vats) {
      target = vats.target;
    },
    start() {
      E(target).create(sensor);
      return sensor;
    },
    drop() {
      return E(target).forget();
    },
  });
}
```

The test program should create these two vats, use `queueToKref` to send `start` to bootstrap, and wait for `c.run()` to settle, then look at the result and extract the kref of `sensor`, then look up that kref in the target vat's c-list to make sure it is still imported. Then it should send `drop` to bootstrap and wait for it to settle, then re-examine the c-list. At this point, if the bug is fixed, the `sensor` kref should be missing, because the target vat should have dropped it along with the rest of `foo`. (Note that you may need to set `virtualObjectCacheSize: 0` and create a dummy object to make sure `foo` is really gone). Of course, run the test before fixing the bug, to make sure it's working.

**Comment by warner** [comment]
time estimate notes: this is one of those issues where it takes 2 minutes to fix, but probably 3 hours to build a proper unit test that fails first and only starts passing once you've fixed the problem properly

**Comment by warner** [comment]
I wrote a test for the first bug (`deleteInternal`), and I was wrong about what was necessary. The approach that worked was to create a collection that held a second collection, which held a Remotable, which closed over an import (`vc1 -> vc2 -> rem1 -> imp1`). The triggering event is a `vc1.delete('vc2')`, which pushes vc2 onto `possiblyDeadSet`. The next BOYD deletes vc2 and thus deletes all the entries of vc2, which calls `removeReachableVref(rem1vref)` (which drops the last RAM reference to `rem1`) but fails to report `doMoreGC` back to the loop. Since nothing else was added to `possiblyDeadSet` or `possiblyRetiredSet`, and `doMoreGC` is `false`, the loop stops, leaving `rem1` UNREACHABLE but not yet COLLECTED. If the bug were fixed, `doMoreGC` would have been `true`, and we'd take another pass through the `scanDeadObjects` loop, allowing `rem1 to become FINALIZED on the same crank, dropping the `imp1` import a crank earlier.
